### ISSUE TITLE
Support custom CXX types and ffi definitions

### ISSUE THREAD

**Issue by ahayzen-kdab** [KDAB/cxx-qt#127]
We would like to support CXX types that the developer has defined themselves to be used with CXX-Qt.

This could look like the following to support a custom type.

```rust
#[make_qobject]
mod my_object {
    // here the developer can describe their own FFI types
    // this will be combined with the generated FFI from CXX-Qt
    mod ffi {
        extern "C++" {
            // any external types the developer will need to include the headers (like with normal CXX)
            // from our CXX-Qt generated QObject header we can include either the CXX header or these headers
            include!("path/to/another.h");

            // developers can define custom types to be used with CXX like normal
            type MyType; // opaque type
            type MyTrivial = rust::my_trivial;
        }

        extern "Rust" {
            // developers can define functions as normal, these need to be implemented in the parent module
            fn f(x: &MyType) -> Box<MyType>;

            // developers can then define methods on the RustObj like a normal CXX method
            // they should mark any that are invokable as invokable here as well, so that we can match them easily
            //
            // note that if opaque types or CppObj are used as types then we need to replace this with a wrapper as usual
            //
            // then we will need to use the same logic as CXX to change Pin<&mut T> -> *T, &T -> const T&, T -> T for the C++ side.
            #[invokable]
            fn custom_type(&self: RustObj, cpp: &mut CppObj, my_type: &MyType, opaque: QColor, trivial: Pin<&mut MyTrivial>, ref: &MyTrivial, value: MyTrivial);

            // This could be the generate wrapper  for the above definition
            fn custom_type_wrapper(&self: RustObj, cpp: Pin<&mut MyObject>, my_type: &MyType, opaque: &QColor, trivial: Pin<&mut MyTrivial>, ref: &MyTrivial, value: MyTrivial);

        }
    }

    #[derive(Default)]
    pub struct Data {
        number: i32,
        string: String,
    }

    #[derive(Default)]
    struct RustObj;

    use cxx_qt_lib::QString;

    // implementation of the function defined in the custom CXX bridge
    fn f(x: &ffi::MyType) -> Box<ffi::MyType> {

    }

    impl RustObj {
        fn private_method() {}

        #[invokable]
        fn custom_type(&self, cpp: &mut CppObj, my_type: &ffi::MyType, opaque: QColor, trivial: Pin<&mut rust::my_trivial>, ref: &rust::my_trivial, value: rust::my_trivial) {
            cpp.set_number(cpp.number() + 1);
        }

        #[invokable]
        fn increment_number(&self, cpp: &mut CppObj) {
            cpp.set_number(cpp.number() + 1);
        }

        #[invokable]
        fn say_hi(&self, string: &QString, number: i32) {
            let s: String = string.into();
            println!("Hi from Rust! String is '{}' and number is {}", s, number);
        }
    }
}
```

**Comment by ahayzen-kdab** [comment]
This could allow for defining methods that are available from the Qt base class as well as custom classes.

Eg

```rust
mod my_object {
    mod ffi {
        extern "C++" {
            fn blockSignals(self: Pin<&mut MyObject>, bool block) -> bool;
            fn signalsBlocked(self: &MyObject) -> bool;
        }
   ...
```

**Comment by ahayzen-kdab** [comment]
This combined with being able to choose a base class in the macro (likely as part of #22 ) could allow for almost manually "implementing" a list model by implementing the correct methods.

```rust
#[cxx_qt(QAbstractListModel)]
mod my_model {
  mod ffi {
    extern "C++" {
      fn beginResetModel(self: Pin<&mut MyModel>);
      fn endResetModel(self: Pin<&mut MyModel>);
    }
  }

  #[derive(Default)]
  struct RustObj {
    items: Vec<T>
  }

  impl RustObj {
    #[invokable]  // we need to mark as override here too ?
    fn data(&self, role: Role) -> QVariant {
        ...
    }

    #[invokable]  // we need to mark as override here too?
    fn role_names(&self) -> QHash<int, String> {  // we don't have QHash or QByteArray
        ...
    }

    #[invokable]
    fn load_entries(&mut self) {
       self.beginResetModel();
       ..
       self.endResetModel();
    }
  }
}

**Comment by jason-yau** [comment]
When will these be supported? And I hope the `Data` struct support cxx custom types.

**Comment by ahayzen-kdab** [comment]
Hopefully at some point in the near future, we are currently discussing a new API which is closer to being a superset of CXX. And we have discussions around how properties might change to being normal fields on the RustObj struct rather than a separate Data struct. We'll write this up soon as a github discussion and start prototyping what this will look like over the next few weeks.

**Comment by Be-ing** [comment]
The proposals have been published as [GitHub Discussions](https://github.com/KDAB/cxx-qt/discussions). @jason-yau if you have any thoughts on those, your input would be appreciated.

**Comment by ahayzen-kdab** [comment]
This is now implemented in `main` (and will be part of 0.4), you can add CXX extern blocks and use those types within properties, invokables, signals.
